python/ray/train/huggingface/accelerate/accelerate_trainer.py

```diff
@@ -0,0 +1,434 @@
+
+        if Version(accelerate.__version__) < Version("0.17.0.dev0"):
+            raise RuntimeError(
+                "AccelerateTrainer requires accelerate>=0.17.0, "
+                f"got {accelerate.__version__}"
+            )
+
+        self.accelerate_config = accelerate_config
+        if isinstance(self.accelerate_config, AccelerateConfigWrapper):
+            self._accelerate_config_raw = self.accelerate_config.config_raw
+            self._deepspeed_config_file_raw = (
+                self.accelerate_config.deepspeed_config_raw
+            )
+            train_loop_config = self._create_train_loop_per_worker(
+                train_loop_per_worker,
+                train_loop_config,
+                self._accelerate_config_raw,
+                self._deepspeed_config_file_raw,
+            )
+        else:
+            (
+                self._accelerate_config_raw,
+                self._deepspeed_config_file_raw,
+            ) = load_accelerate_config(self.accelerate_config)
+        super().__init__(
+            train_loop_per_worker=_accelerate_train_loop_per_worker,
+            train_loop_config=train_loop_config,
+            torch_config=torch_config,
+            scaling_config=scaling_config,
+            dataset_config=dataset_config,
+            run_config=run_config,
+            datasets=datasets,
+            preprocessor=preprocessor,
+            resume_from_checkpoint=resume_from_checkpoint,
+        )
+
+    def as_trainable(self) -> Type["Trainable"]:
+        # We want to load the config when the Trainer is first instantiated,
+        # and share the contents with the Trainables (which may be on different)
+        # nodes
+        old_accelerate_config = self._param_dict.get("accelerate_config", None)
+        self._param_dict["accelerate_config"] = AccelerateConfigWrapper(
+            self._accelerate_config_raw, self._deepspeed_config_file_raw
+        )
+        try:
+            ret = super().as_trainable()
+        finally:
+            self._param_dict["accelerate_config"] = old_accelerate_config
+        return ret
+
+    @classmethod
+    def _create_train_loop_per_worker(
+        cls,
+        train_loop_per_worker: Union[Callable[[], None], Callable[[Dict], None]],
+        train_loop_per_worker_config,
+        accelerate_config_raw: str,
+        deepspeed_config_file_raw: str,
+    ):
+        """Wrap around train_loop_per_worker to set necessary Accelerate env vars."""
+        train_loop_per_worker_config = (
+            train_loop_per_worker_config.copy() if train_loop_per_worker_config else {}
+        )
+        for key in train_loop_per_worker_config:
+            if key in RESERVED_KEYS:
+                raise ValueError(
+                    f"'{key}' is a reserved key in `train_loop_per_worker_config`."
+                )
+        if train_loop_per_worker:
+            train_loop_per_worker_config[
+                TRAIN_LOOP_PER_WORKER_KEY
+            ] = train_loop_per_worker
```

Comment:
Is there a cleaner way such that we don't have to do this... 

Preferably we can create the final `train_loop_per_worker` from the input `train_loop_per_worker` and accelerate configs, without using `config` as a container.

Reply:
sure - that was actually the approach I took first!